File: frontend/src/types/facility.ts
/** Facility (datacenter / rack / rack elevation) type definitions */

export interface Rack {
  id: number
  datacenter: string | null
  name: string
  u_height: number
  room: string | null
  row_label: string | null
  col_label: string | null
  description: string | null
}

export interface RackCreatePayload {
  datacenter?: string | null
  name: string
  u_height?: number
  room?: string | null
  row_label?: string | null
  col_label?: string | null
  description?: string | null
}

export interface RackUpdatePayload {
  datacenter?: string | null
  name?: string
  u_height?: number
  room?: string | null
  row_label?: string | null
  col_label?: string | null
  description?: string | null
}

export interface ConnectedSwitch {
  id: number
  asset_no: string
  hostname: string | null
  asset_type: string
}

export interface DeviceInterface {
  id: number
  name: string | null
  mac_address: string | null
  ip_address: string | null
  role: string
  vlan_id: number | null
  bond_master: string | null
  is_primary: boolean
  connected_asset_id: number | null
  connected_port: string | null
  connected_switch: ConnectedSwitch | null
  source: 'port' | 'asset'
}

export interface ElevationDevice {
  id: number
  asset_no: string
  hostname: string | null
  asset_type: string
  importance: string
  status: string
  u_start: number | null
  u_height: number | null
  interfaces: DeviceInterface[]
}

export interface RackWarning {
  code: 'out_of_range' | 'overlap' | 'missing_u_start'
  message: string
  asset_id: number | null
}

export interface RackElevation extends Rack {
  devices: ElevationDevice[]
  warnings: RackWarning[]
}

export interface RackSummary {
  id: number
  name: string
  u_height: number
  device_count: number
  used_u: number
  utilization: number
}

export interface UnlocatedAsset {
  id: number
  asset_no: string
  hostname: string | null
  ip_address: string
  reason: 'no_rack' | 'no_u_start'
}

export interface DatacenterRacks {
  datacenter: string
  racks: RackSummary[]
  unlocated: UnlocatedAsset[]
}

export interface DatacenterSummary {
  datacenter: string
  rack_count: number
  device_count: number
  used_u: number
  total_u: number
  utilization: number
  unlocated_count: number
}

File: backend/app/schemas/facility.py
"""Facility (datacenter / rack / rack elevation) schemas."""
from pydantic import BaseModel, Field


# ── Rack CRUD ────────────────────────────────────────────────────

class RackCreate(BaseModel):
    datacenter: str | None = Field(None, max_length=128)
    name: str = Field(..., max_length=64)
    u_height: int = Field(42, ge=1, le=52)
    room: str | None = Field(None, max_length=128)
    row_label: str | None = Field(None, max_length=64)
    col_label: str | None = Field(None, max_length=64)
    description: str | None = None


class RackUpdate(BaseModel):
    datacenter: str | None = Field(None, max_length=128)
    name: str | None = Field(None, max_length=64)
    u_height: int | None = Field(None, ge=1, le=52)
    room: str | None = Field(None, max_length=128)
    row_label: str | None = Field(None, max_length=64)
    col_label: str | None = Field(None, max_length=64)
    description: str | None = None


class RackRead(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    datacenter: str | None
    name: str
    u_height: int
    room: str | None
    row_label: str | None
    col_label: str | None
    description: str | None


# ── Aggregated views ─────────────────────────────────────────────

class ConnectedSwitch(BaseModel):
    id: int
    asset_no: str
    hostname: str | None
    asset_type: str


class DeviceInterface(BaseModel):
    id: int
    name: str | None
    mac_address: str | None
    ip_address: str | None
    role: str
    vlan_id: int | None
    bond_master: str | None
    is_primary: bool
    connected_asset_id: int | None
    connected_port: str | None
    connected_switch: ConnectedSwitch | None = None
    source: str = "port"  # port = port-level cabling; asset = device-level fallback


class ElevationDevice(BaseModel):
    id: int
    asset_no: str
    hostname: str | None
    asset_type: str
    importance: str
    status: str
    u_start: int | None
    u_height: int | None
    interfaces: list[DeviceInterface] = []


class RackWarning(BaseModel):
    code: str            # out_of_range / overlap / missing_u_start
    message: str
    asset_id: int | None = None


class RackElevation(RackRead):
    devices: list[ElevationDevice] = []
    warnings: list[RackWarning] = []


class RackSummary(BaseModel):
    id: int
    name: str
    u_height: int
    device_count: int
    used_u: int
    utilization: float


class UnlocatedAsset(BaseModel):
    id: int
    asset_no: str
    hostname: str | None
    ip_address: str
    reason: str          # no_rack / no_u_start


class DatacenterRacks(BaseModel):
    datacenter: str
    racks: list[RackSummary]
    unlocated: list[UnlocatedAsset] = []


class DatacenterSummary(BaseModel):
    datacenter: str
    rack_count: int
    device_count: int
    used_u: int
    total_u: int
    utilization: float
    unlocated_count: int
